examples/call-runtime/lib.rs

```diff
@@ -0,0 +1,222 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink::primitives::AccountId;
+use sp_runtime::MultiAddress;
+
+/// A part of the runtime dispatchable API.
+///
+/// For now, `ink!` doesn't provide any support for exposing the real `RuntimeCall` enum, which
+/// fully describes the composed API of all the pallets present in runtime. Hence, in order to use
+/// `call-runtime` functionality, we have to provide at least a partial object, which correctly
+/// encodes the target extrinsic.
+///
+/// You can investigate the full `RuntimeCall` definition by either expanding `construct_runtime!`
+/// macro application or by using secondary tools for reading chain metadata, like `subxt`.
+#[derive(scale::Encode)]
+enum RuntimeCall {
+    // This index can be found by investigating runtime configuration. You can check the pallet
+    // order inside `construct_runtime!` block and read the position of your pallet (0-based).
+    //
+    // See also
+    // [stack question](https://substrate.stackexchange.com/questions/778/how-to-get-pallet-index-u8-of-a-pallet-in-runtime).
+    #[codec(index = 4)]
+    Balances(BalancesCall),
+}
+
+#[derive(scale::Encode)]
+enum BalancesCall {
+    // This index can be found by investigating the pallet dispatchable API. In your pallet code,
+    // look for `#[pallet::call]` section and check `#[pallet::call_index(x)]` attribute of the
+    // call. If these attributes are missing, use source-code order (0-based).
+    #[codec(index = 0)]
+    Transfer {
+        dest: MultiAddress<AccountId, ()>,
+        #[codec(compact)]
+        value: u128,
+    },
+}
+
+#[ink::contract]
+mod runtime_call {
+    use crate::{
+        BalancesCall,
+        RuntimeCall,
+    };
+
+    /// A trivial contract with a single message, that uses `call-runtime` API for performing
+    /// native token transfer.
+    #[ink(storage)]
+    #[derive(Default)]
+    pub struct RuntimeCaller;
+
+    impl RuntimeCaller {
+        /// The constructor is `payable`, so that during instantiation it can be given some tokens
+        /// that will be further transferred with `transfer_through_runtime` message.
+        #[ink(constructor, payable)]
+        pub fn new() -> Self {
+            Default::default()
+        }
+
+        /// Tries to transfer `value` from the contract's balance to `receiver`.
+        ///
+        /// Fails if:
+        ///  - called in the off-chain environment
+        ///  - the chain doesn't allow `call-runtime` API (`UnsafeUnstableInterface` is turned off)
+        ///  - the chain forbids contracts to call `Balances::transfer` (`CallFilter` is too
+        ///    restrictive)
+        ///  - after the transfer, `receiver` doesn't have at least existential deposit  
+        ///  - the contract doesn't have enough balance
+        #[ink(message)]
+        pub fn transfer_through_runtime(
+            &mut self,
+            receiver: AccountId,
+            value: Balance,
+        ) -> Result<(), ()> {
```

Comment:
I'm surprised Clippy hasn't complained about this.

Usually it says to use structs here.

E.g 

```rust
struct CallRuntimeError;
fn foo() -> Result<(), CallRuntimeError>;
```

We should try and get the error from the contracts pallet here, and that way in the failure test we can also check that we get the correct error, e.g `assert!(matches!(res, CallRuntimeFailed))`

Reply:
Why should it? `()` is a correct type for error. I was thinking indeed about introducing custom error, but the only reasonable thing that can be returned is `CallRuntimeFailed`, so there is no much point in testing it. Also, it requires another dependency (`ink_env`)

Reply:
Here's the [Clippy lint](https://rust-lang.github.io/rust-clippy/master/index.html#result_unit_err).

>  but the only reasonable thing that can be returned is CallRuntimeFailed, so there is no much point in testing it

We want to ensure that indeed it does get returned and not that we error out with something like a generic trap for example.

You can access `ink_env` through `ink::env` now, so that's not an issue.

Here's what I recommend:

```rust
#[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
pub enum RuntimeError {
    CallRuntimeFailed,
}
                                                                                   
impl From<ink::env::Error> for RuntimeError {
    fn from(e: ink::env::Error) -> Self {
        match e {
            ink::env::Error::CallRuntimeFailed => RuntimeError::CallRuntimeFailed,
            _ => panic!(),
        }
    }
}
```

And then in the test we check for that.

Reply:
the problem is that `client.call` will return an opaque error type with `subxt::error::DispatchError` in the end, so I can do nothing more than `assert!(matches!(call_res, Err(ink_e2e::Error::CallExtrinsic(_))));`, can I?

Reply:
Actually, digging in a bit more that makes sense. We have four cases here:

1. `call_runtime` is enabled and the `Call` is valid

We expect to get an `Ok(())` from our message

2. `call_runtime` is enabled and the `Call` is valid but returns an error (e.g an invalid transfer amount)

We expect to get an `Err(CallRuntimeReturnedError)`

3. `call_runtime` is enabled and the `Call` is invalid (e.g non-existent call index)

We expect to get an error from the contracts pallet during dispatch (for the non-existent example I was getting a decoding error)

4. `call_runtime` is not enabled

Dispatch should fail since we're being blocked by the filter.

We still need to add tests for (2) and (3).

For your question about errors, for the calls that we expect should fail you should use the `call_dry_run` method instead and dig into those errors. For (4) the `CallExtrinsic` error seems correct.

Reply:
missing testcases were added; I needed a new artificial failing message and unfortunately it has to be visible in any context - I tried adding `#[cfg(all(test, feature = "e2e-tests"))]` attribute there, but it generates some compilation errors

Reply:
I don't know what you mean by the "artificial error message". New set of tests looks like though

Reply:
that this message is for failure tests only; anyway, glad that we are done here